subsys/storage/storage_area/storage_area.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2024 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <errno.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/storage/storage_area/storage_area.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(storage_area, CONFIG_STORAGE_AREA_LOG_LEVEL);
+
+static bool sa_range_valid(const struct storage_area *area, size_t start,
+			   size_t len)
+{
+	const size_t asize = area->erase_size * area->erase_blocks;
+
+	if ((asize < len) || ((asize - len) < start)) {
+		LOG_DBG("Invalid range");
+		return false;
+	}
+
+	return true;
+}
+
+static size_t sa_iovec_size(const struct storage_area_iovec *iovec,
+			    size_t iovcnt)
+{
+	size_t rv = 0U;
+
+	for (size_t i = 0U; i < iovcnt; i++) {
+		rv += iovec[i].len;
+	}
+
+	return rv;
+}
+
+int storage_area_readv(const struct storage_area *area, size_t start,
+		       const struct storage_area_iovec *iovec, size_t iovcnt)
+{
+	if ((area == NULL) || (area->api == NULL) || (area->api->readv == NULL)) {
+		return -ENOTSUP;
+	}
```

Comment:
I am not entirely sure on this one. When we give access to these routines through the shell it could be possible to provide wrong values to the routines. It also seems that the use of `__ASSERT_NO_MSG` is discouraged.

Regarding the existence of a storage area without read method I think this is very possible, think of a remote location that only allows writing and not reading.

If you have no objection I would like to keep it as is.